### PR TITLE
Implement on-type-rename for TypeScript

### PR DESCRIPTION
**Pull request by jrieken**
This PR fixes nothing but explores on-type-rename for TS. There is no "proper" tsserver support for this but what I have seems to work surprisingly well.

### CODE REVIEW

File: extensions/typescript-language-features/src/languageFeatures/renameOnType.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import { ClientCapability, ITypeScriptServiceClient } from '../typescriptService';
+import { conditionalRegistration, requireSomeCapability } from '../utils/dependentRegistration';
+import { DocumentSelector } from '../utils/documentSelector';
+import * as typeConverters from '../utils/typeConverters';
+
+
+class TypeScriptOnTypeRenameProvider implements vscode.OnTypeRenameProvider {
+
+	private static enabledKinds = new Set<string>([
+		'let', 'const', 'local var', 'parameter'
```
Comment (mjbvz) [inline extensions/typescript-language-features/src/languageFeatures/renameOnType.ts:16]: [nit] Use the [`protocol.const` static variables](https://github.com/microsoft/vscode/blob/9d6054c99d63b934148051f1fb5ae22e926cc169/extensions/typescript-language-features/src/protocol.const.ts#L6) here (this lets us use find all references on them)

**Review by mjbvz — APPROVED** [review]
Neat! Change looks good to me overall

Should we ask TS for a proper API? Besides having to make two calls, are there any major issues with the current approach that TS could solve?

**Comment by mjbvz** [comment]
One other question: does this work for renaming JSX elements? This was a highly upvoted issue: #85707

**Review by gbumanzordev — APPROVED** [review]

**Review by ghost — CHANGES_REQUESTED** [review]
>  lllll`

File: extensions/typescript-language-features/src/languageFeatures/renameOnType.ts (hunk 1 above)
Comment (jrieken) [reply extensions/typescript-language-features/src/languageFeatures/renameOnType.ts]: check

**Comment by jrieken** [comment]
fyi - I have also added `type parameter` to the list of kinds and I am skipping when there isn't at least 2 places to quick-rename. 

> One other question: does this work for renaming JSX elements? This was a highly upvoted issue: #85707

For JSX it seem that they are reported as `property` which is a little vague since that's also used for ordinary properties of TS/JS types. I don't know enough about JSX to know if that's enough or not but my understanding is that renaming those can potentially cause renamed in other files and that's currently unsupported

**Comment by yume-chan** [comment]
> One other question: does this work for renaming JSX elements? 

Is it proper to call is "renaming"? Say I have two components: `Bar` and `Baz`, what mirror cursor can do is "change this Bar tag to be a Baz tag", not "rename Bar component to Baz", resulting two components both called `Baz`.

Looking at the implementation I think this pr is also doing the later

```ts
let bar = 1;
let baz = 2;
console.log(bar); // change this `bar` to `baz`, I don't want to rename the variable to `baz`, I just want to use `baz` instead
```

**Comment by jrieken** [comment]
> Looking at the implementation I think this pr is also doing the later

Yes - as of today it like f2-rename. Which the name of the feature also implies. What I hear from you is to only rename usages and not the declaration in some cases. Unsure how such cases would be defined? Depending on where the cursor location is?

**Comment by yume-chan** [comment]
In Visual Studio and C#, after changing the name of something, there will be a code action to let you apply the edit as renaming, it will never do this by its own.

Maybe it's feasible to automatically rename something when cursor is on its declaration, but it's still possible that the user really want to do is replacing the implementation, so they rename the current one (as a backup), and add a new one with same name.

**Comment by jrieken** [comment]
> Maybe it's feasible to automatically rename something when cursor is on its declaration, but it's still possible that the user really want to do is replacing the implementation, 

Sure, sure. That's why this is a feature that's off by default and also when its active you get clear feedback UX and can then cancel via ESC

**Pull request merged**